src/box/memtx_engine.c

```diff
@@ -1049,10 +1053,220 @@ memtx_engine_gc_f(va_list va)
+		memtx_allocator_enter_delayed_free_mode(memtx);
+}
+
+void
+memtx_leave_delayed_free_mode(struct memtx_engine *memtx)
+{
+	assert(memtx->delayed_free_mode > 0);
+	if (--memtx->delayed_free_mode == 0)
+		memtx_allocator_leave_delayed_free_mode(memtx);
+}
+
+#define DECLARE_MEMTX_TUPLE_NEW(type, prefix)					\
+struct tuple *  								\
+type##_##memtx_tuple_new(struct tuple_format *format,				\
+	const char *data, const char *end)					\
+{										\
+	struct memtx_engine *memtx = (struct memtx_engine *)format->engine;	\
+	assert(mp_typeof(*data) == MP_ARRAY);					\
+	struct tuple *tuple = NULL;						\
+	struct region *region = &fiber()->gc;					\
+	size_t region_svp = region_used(region);				\
+	struct field_map_builder builder;					\
+	if (tuple_field_map_create(format, data, true, &builder) != 0)		\
+		goto end;							\
+	uint32_t field_map_size = field_map_build_size(&builder);		\
+	/*
+	 * Data offset is calculated from the begin of the struct
+	 * tuple base, not from memtx_tuple, because the struct
+	 * tuple is not the first field of the memtx_tuple.
+	 */									\
+	uint32_t data_offset = sizeof(struct tuple) + field_map_size;		\
+	if (data_offset > INT16_MAX) {						\
+		/** tuple->data_offset is 15 bits */				\
+		diag_set(ClientError, ER_TUPLE_METADATA_IS_TOO_BIG,		\
+			 data_offset);						\
+		goto end;							\
+	}									\
+										\
+	size_t tuple_len = end - data;						\
+	size_t total = sizeof(struct memtx_tuple) + field_map_size +		\
+			tuple_len;						\
+										\
+	ERROR_INJECT(ERRINJ_TUPLE_ALLOC, {					\
+		diag_set(OutOfMemory, total, "slab allocator", "memtx_tuple");	\
+		goto end;							\
+	});									\
+	if (unlikely(total > memtx->max_tuple_size)) {				\
+		diag_set(ClientError, ER_MEMTX_MAX_TUPLE_SIZE, total);		\
+		error_log(diag_last_error(diag_get()));				\
+		goto end;							\
+	}									\
+										\
+	struct memtx_tuple *memtx_tuple;					\
+	while ((memtx_tuple = prefix##alloc((struct type##_alloc *)		\
+			memtx->alloc, total)) == NULL) {			\
+		bool stop;							\
+		memtx_engine_run_gc(memtx, &stop);				\
+		if (stop)							\
+			break;							\
+	}									\
+	if (memtx_tuple == NULL) {						\
+		diag_set(OutOfMemory, total, "slab allocator", "memtx_tuple");	\
+		goto end;							\
+	}									\
+	tuple = &memtx_tuple->base;						\
+	tuple->refs = 0;							\
+	memtx_tuple->version = memtx->snapshot_version;				\
+	assert(tuple_len <= UINT32_MAX); /* bsize is UINT32_MAX */		\
+	tuple->bsize = tuple_len;						\
+	tuple->format_id = tuple_format_id(format);				\
+	tuple_format_ref(format);						\
+	tuple->data_offset = data_offset;					\
+	tuple->is_dirty = false;						\
+	char *raw = (char *) tuple + tuple->data_offset;			\
+	field_map_build(&builder, raw - field_map_size);			\
+	memcpy(raw, data, tuple_len);						\
+	say_debug("%s(%zu) = %p", __func__, tuple_len, memtx_tuple);		\
+end:										\
+	region_truncate(region, region_svp);					\
+	return tuple;								\
+}
```

Comment:
Instead of this macro-hell (more than 100 lines of macros!!!) I would suggest to turn this file into .cpp, and try templates maybe. If possible.

Reply:
I significantly reduced the number of macros and encapsulated them in the allocator.h. Also i done this work without macros in mechanik20051988/gh-5419-choose-allocator-for-memtx-cpp using templates.